core/src/main/java/org/apache/calcite/rel/rules/FilterExtractInnerJoinRule.java

```diff
@@ -0,0 +1,213 @@
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexLiteral;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexSubQuery;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.tools.RelBuilder;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Stack;
+import java.util.stream.Collectors;
+
+/**
+ * Planner rule that matches an {@link org.apache.calcite.rel.core.Filter}
+ * on a {@link org.apache.calcite.rel.core.Join} and removes the join
+ * predicates from the filter conditions and put them on Join if possible.
+ *
+ * <p>For instance,
+ *
+ * <blockquote>
+ * <pre>select e.employee_id, e.name
+ * from employee as e, department as d
+ * where e.department_id = d.department_id and d.department_id > 5</pre></blockquote>
+ *
+ * <p>becomes
+ *
+ * <blockquote>
+ * <pre>select e.employee_id, e.name
+ * from employee as e
+ * INNER JOIN department as d
+ * ON e.department_id = d.department_id
+ * WHERE d.department_id > 5</pre></blockquote>
+ *
+ * @see CoreRules#FILTER_EXTRACT_INNER_JOIN_RULE
+ */
+public class FilterExtractInnerJoinRule
+    extends RelRule<FilterExtractInnerJoinRule.Config>
+    implements TransformationRule {
+
+  protected FilterExtractInnerJoinRule(Config config) {
+    super(config);
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Filter filter = call.rel(0);
+    final Join join = call.rel(1);
+    RelBuilder builder = call.builder();
+
+    if (!isCrossJoin(join, builder)
+        || isFilterWithCompositeLogicalConditions(filter.getCondition())) {
+      return;
+    }
+
+    Stack<Map.Entry<RelNode, Integer>> stackForTablesWithEndIndexes = new Stack<>();
+    List<RexNode> allConditions = new ArrayList<>();
+    populateStackWithEndIndexesForTables(join, stackForTablesWithEndIndexes, allConditions);
+    if (((RexCall) filter.getCondition()).getOperands().stream().allMatch(operand ->
+        operand instanceof RexCall && ((RexCall) operand).operands.size() > 1)) {
+      allConditions.addAll(((RexCall) filter.getCondition()).getOperands());
+    } else {
+      allConditions.add(filter.getCondition());
+    }
+
+    final RelNode modifiedJoinClauseWithWhereClause =
+        moveConditionsFromWhereClauseToJoinOnClause(
+            allConditions, stackForTablesWithEndIndexes, builder);
+
+    call.transformTo(modifiedJoinClauseWithWhereClause);
+  }
+
+  private static boolean isCrossJoin(Join join, RelBuilder builder) {
+    return join.getJoinType().equals(JoinRelType.INNER)
+        && builder.literal(true).equals(join.getCondition());
+  }
+
+  /** This method checks whether filter conditions have both AND & OR in it.*/
+  private static boolean isFilterWithCompositeLogicalConditions(RexNode condition) {
```

Comment:
In each line of code `condition` is casted  to `RexCal`l, can we change the argument of  to `isFilterWithCompositeLogicalConditions(RexCall call) ` ?

Reply:
Done!